library/bdBank/Deferred/User.php

```diff
@@ -69,4 +84,35 @@ protected function _fetchSum($formula, $where)
 
         return bdBank_Helper_Number::add($liveValue, $archiveValue);
     }
+
+    protected function _fetchAdjustedSum($userId, $direction)
+    {
+        if ($direction == 'incoming') {
+            $userIdCol = 'to_user_id';
+        } elseif ($direction == 'outgoing') {
+            $userIdCol = 'from_user_id';
+        } else {
+            throw new XenForo_Exception('Unsupported direction');
+        }
+        $db = XenForo_Application::getDb();
+
+        $totalAdjustedAmount = $db->fetchOne("
+            SELECT SUM(adj.amount)
+            FROM xf_bdbank_transaction_adjustment adj
+                LEFT JOIN xf_bdbank_transaction t
+                    ON adj.comment = t.comment
+                        AND t.reversed = 0
+                LEFT JOIN xf_bdbank_archive a
+                    ON adj.comment = a.comment 
+            WHERE t.$userIdCol = $userId
```

Comment:
This query may get quite expensive. Have you tested with enough data?

Reply:
hmm all `comment`, `from_user_id` and `to_user_id` columns are indexed (except `archive.from_user_id`, don't know why?).  
Also this operation is rarely executed so I think it's fine?  

Besides, not sure the following is better:
```sql
SELECT SUM(adj.amount)
FROM xf_bdbank_transaction_adjustment adj
WHERE adj.comment IN (
    SELECT t.comment
    FROM xf_bdbank_transaction t
    WHERE t.$userIdCol = $userId
      AND t.reversed = 0
    UNION ALL
    SELECT a.comment
    FROM xf_bdbank_archive a
    WHERE a.$userIdCol = $userId
)
```

Reply:
I think the problematic table is adjustment. It doesn't have any index so when you left join it will scan all adjustments. We will have millions of those -> it will be slow.

The sub query in WHERE looks better I think. I do recall some issues with sub query though, maybe we need to benchmark this with real data. How many rows do you have?

Reply:
I added an index on adjustment.comment anh.  
Let me do some benchmarkings